File: PathSum-II.java
// Time Complexity :O(n)
// Space Complexity : O(n) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
            List<List<Integer>> result;

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        this.result = new ArrayList<>();
        helper(root,0,targetSum, new ArrayList<>());
        return result;
    }
    
   private void helper(TreeNode root,int currentSum,int target, List<Integer> path){
       
   //Base
    if(root == null) 
        return;
       
    currentSum +=   root.val;
    //action
    path.add(root.val);
       
    //pre
    if(currentSum == target && root.left == null && root.right == null)    
       result.add(new ArrayList<>(path));
    //recursion
    helper(root.left, currentSum,target, path);  
    //st.pop()
    helper(root.right, currentSum,target, path);
    //st.pop()
    //backtrack
    path.remove(path.size()-1);  
     
      
       
   }

    
    
    
    
    
}

File: pathSumII.cpp
//time complexity: O(n)
//space complexity:O(n)
//approach: recursion
//solved on leetcode? yes
//problem faced? no

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>>res;
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        if(root==NULL)
            return res;
        vector<int>path;
        helper(root,sum,path);
        return res;
    }
    void helper(TreeNode* root,  int sum, vector<int>path)
    {
        if(root==NULL)
            return;
        if(root->left==NULL && root->right==NULL)
        {
            if(sum-root->val==0)
            {
                path.push_back(root->val);
                res.push_back(path);
            }
            return;
        }
        path.push_back(root->val);
        helper(root->left,sum-root->val,path);
        helper(root->right,sum-root->val,path);
        return;
    }
};